flashlight/fl/tensor/backend/jit/opt/passes/ScalarFolding.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "flashlight/fl/tensor/backend/jit/opt/passes/ScalarFolding.h"
+
+#include "flashlight/fl/tensor/backend/jit/ir/BinaryNode.h"
+#include "flashlight/fl/tensor/backend/jit/ir/ScalarNode.h"
+
+namespace fl {
+
+namespace {
+
+template <typename T>
+T foldScalars(const T lhs, const T rhs, const BinaryOp op) {
+  switch (op) {
+    case BinaryOp::Add:
+      return lhs + rhs;
+    case BinaryOp::Sub:
+      return lhs - rhs;
+    case BinaryOp::Mul:
+      return lhs * rhs;
+    case BinaryOp::Div:
+      return lhs / rhs;
+  }
+  throw std::runtime_error("[foldScalars] Unknown binary operation type");
+}
+
+template <typename T>
+ScalarNode* foldScalarNodes(
+    const ScalarNode& lhs,
+    const ScalarNode& rhs,
+    const BinaryOp op,
+    const dtype type) {
+  T lhsVal = lhs.scalar<T>();
+  T rhsVal = rhs.scalar<T>();
+  T resVal = foldScalars(lhsVal, rhsVal, op);
+  return ScalarNode::create(Shape(lhs.shape()), type, resVal);
+}
+
+ScalarNode* foldScalarNodes(
+    const ScalarNode& lhs,
+    const ScalarNode& rhs,
+    const BinaryOp op,
+    const dtype type) {
+  switch (type) {
```

Comment:
No better idea at the moment, but I wish there were a better way to trait away what will invariably be many runtime --> compile time type switches

Reply:
Yeah I've been observing and thinking about this.

I have a macro-based idea in mind, and I'll send a proof-of-concept PR at some point.